launch_testing/launch_testing/loader.py

```diff
@@ -17,9 +17,14 @@
 import itertools
 import unittest
 
+import launch.logging
+
 from .actions import ReadyToTest
 
 
+_logger = launch.logging.get_logger(__name__)
```

Comment:
@pbaughman meta: I wonder if we shouldn't use the `warnings` module instead.

Reply:
I'm not sure I've used that module before, so I can't speak to the pros and cons.  You're referring [to this](https://docs.python.org/3/library/warnings.html) right?

Reply:
Right.

Reply:
@hidmic Using the warnings module changes

```
test_count_to_four (good_proc_launch_test.TestGoodProcess) ... [WARNING] [launch_testing.loader]: Automatically adding attributes like self.proc_outputto the test class will be deprecated in a future release.  Instead, add proc_output to the test method argument list to access the test object you need
```

To 

```
test_count_to_four (good_proc_launch_test.TestGoodProcess) ... /launch/install/launch_testing/lib/python3.6/site-packages/launch_testing/loader.py:251: FutureWarning: Automatically adding attribut
es like self.proc_output to the test class will be deprecated in a future release.  Instead, add proc_output to the test method argument list to access the test object you need
  FutureWarning,
```

It tries to print 1 line of source code at the end - that's why it says "FutureWarning"  I'm not sure how to get it to stop doing that, but I also didn't look very hard.

It doesn't seem like using the 'warnings' module is a clear winner here

Reply:
I don't know how you're using the module. I think that using the `stacklevel` parameter appropriately you may get to refer to the exact line that triggered the warning (which is cool IMO).

Reply:
@hidmic It's already pointing to the exact line that triggered the warning.  The source code is:

```
warnings.warn(                                                                         
    'Passing ready_fn as an argument to generate_test_description will '               
    'be deprecated in a future release.  Include a launch_testing.actions.ReadyToTest '
    'action in the LaunchDescription instead.',                                        
    FutureWarning,                                                                     
)                                                                                      
```

That's where the somewhat useless "FutureWarning," comes from in the printed message.

I can play with the stacklevel bit.  When warning about 'self.proc_info' and 'self.proc_output' I can get it to point to the test file that the user wrote:

```
test_full_output (good_proc_launch_test.TestProcessOutput) ... /launch/launch_testing/test/launch_testing/examples/good_proc_launch_test.py:81: FutureWarning: Automatically adding attributes like self.proc_output to the test class will be deprecated in a future release.  Instead, add proc_output to the test method argument list to access the test object you need
  with assertSequentialStdout(self.proc_output, dut_process) as cm:
```

When warning about using the ready_fn I doubt I can get it to show part of the generate_test_description function because it's long-gone from the stack by the time raise this warning.

Then options as I see them are:
1) Use warnings.warn for both warnings, but the one about the ready_fn has a useless extra "FutureWarning," printed at the end which is confusing
2) Use warnings.warn for one warning, but not the other - now we have two different formatted warnings, but at least one of them prints the file name and the line that it came from
3) Use logger warnings for both.  They're formatted the same, and work the same in the code but we don't see what line a warning for self.proc_info or self.proc_output comes from

Reply:
> When warning about using the ready_fn I doubt I can get it to show part of the generate_test_description function because it's long-gone from the stack by the time raise this warning.

True that. Hmm, what if we warn them as soon as we see that `generate_test_description` takes a `ready_fn` argument? To make it extra robust, we should also either ban a user-provided `ready_fn` parameter or stay silent and just use the user's if one's actually there.

Reply:
If we use `warnings`, then we see the warning messages if the tests happen to be run with `pytest` (instead of `ros2 test`). I tried locally and do not see launch logs if I use `pytest`.

Reply:
Is this because of the way the pytest wrapper for launch tests captures all stdout and only shows it on failure?  IE, if you run with `-s` do you see the warning output?

Reply:
I do see the warning with `-s`, albeit buried in the rest of the test output. If we use the `warnings` module then we also see warning messages in the pytest summary. Using the `warnings` module also lets users to handle the warnings differently if they want, for example ignoring them or treating them as errors. Perhaps this is also configurable with the launch logger; I don't know.

Reply:
Yeah, I really dislike how the pytest wrapper combines as many tests as there are in a file into one mega-test.  It's quite difficult to interpret the output.  I can switch this to the 'warnings' module as long as we don't go around and around about the message that's printed.  I didn't take that route originally because I couldn't figure out how to get it to display a warning without a bunch of meaningless (to the end user) extra stuff

Reply:
I think if you use the warnings module with the default warning (UserWarning), then I don't think it displays "extra stuff". Even if it is ugly, I think it gets the message across. IMO, the warnings module is more visible for users compared to the launch logger warning. If someone else wants to weigh in, I'm not going to die on this hill. @hidmic ?

Reply:
Had to parse the discussion again, it's been a while.

To let the user know a feature is deprecated, I'd certainly stick to the `warnings` module. Not only for `launch`, but for every other ROS 2 core Python package out there too. @pbaughman I still think you may be able to get better output for `ready_fn` deprecation warnings if you do so when you bind them to the test function instead.

Reply:
And about `pytest` aggregating all output in a rather confusing way, I agree. That's because there's a clear boundary between `launch_testing` and `pytest` (see [here](https://github.com/ros2/launch/blob/a5ec2a46595e7b711a1ffc70fe2b2dfd966aac9c/launch_testing/launch_testing/pytest/hooks.py#L42-L61)), and thus an entire launch test (with all its test cases) is a single `pytest`. To get fine grained output we need deeper integration. Turning `launch_testing` into an extension of `pytest` has been and is a path to achieve that.